source/s3_meta_request.c

```diff
@@ -2320,51 +2395,94 @@ struct aws_s3_meta_request_poll_write_result aws_s3_meta_request_poll_write(
         meta_request->synced_data.async_write.waker = waker;
         meta_request->synced_data.async_write.waker_user_data = user_data;
         result.is_pending = true;
-
     } else {
         /* write call is OK */
 
         /* If we don't already have a buffer, grab one from the pool. */
         if (meta_request->synced_data.async_write.buffered_data_ticket == NULL) {
-            /* NOTE: we acquire a forced-buffer because there's a risk of deadlock if we
-             * waited for a normal ticket reservation, respecting the pool's memory limit.
-             * (See "test_s3_many_async_uploads_without_data" for description of deadlock scenario) */
-            meta_request->synced_data.async_write.buffered_data = aws_s3_buffer_pool_acquire_forced_buffer(
-                meta_request->client->buffer_pool,
-                meta_request->part_size,
-                &meta_request->synced_data.async_write.buffered_data_ticket /*out_new_ticket*/);
-        }
 
-        /* Copy as much data as we can into the buffer */
-        struct aws_byte_cursor processed_data =
-            aws_byte_buf_write_to_capacity(&meta_request->synced_data.async_write.buffered_data, &data);
+            if (meta_request->synced_data.async_write.buffered_data_ticket == NULL &&
+                meta_request->synced_data.async_write.buffered_ticket_future == NULL) {
+                /* NOTE: we acquire a forced-buffer because there's a risk of deadlock if we
+                 * waited for a normal ticket reservation, respecting the pool's memory limit.
+                 * (See "test_s3_many_async_uploads_without_data" for description of deadlock scenario) */
 
-        /* Don't store EOF unless we've consumed all data */
-        if ((data.len == 0) && eof) {
-            meta_request->synced_data.async_write.eof = true;
-        }
+                struct aws_s3_buffer_pool_reserve_meta meta = {
+                    .size = meta_request->part_size,
+                    .can_block = true,
+                    .meta_request = meta_request,
+                    .client = meta_request->client};
 
```

Comment:
The debug log statements `AWS_LOGF_DEBUG(0, "here")` and `AWS_LOGF_DEBUG(0, "not here")` should be removed or replaced with more meaningful messages. These appear to be temporary debug statements.